### PR TITLE
Update the Halogen guide for Halogen 5

### LINKED ISSUE #528: Bring guide to completion
I'd really like to see the halogen guide completed, because it's by far the easiest way to learn about halogen.

I can put in some time next week to write about lifecycle methods, provided that somebody who really knows what's going on plays editor. I cannot be much help with event sources, but I think this chapter is badly needed as well.

Anyone willing to team up and commit to this?

### LINKED ISSUE #588: Documentation is not up to date for master branch
https://github.com/slamdata/purescript-halogen/blob/master/docs/5%20-%20Parent%20and%20child%20components.md

Is misleading as the references to `ParentHTML` and `ParentDSL` give the impression that there is a difference between parent and child components.

### LINKED ISSUE #633: Help offered for next release
@garyb I was wondering if there is anything I could do to help make progress on Halogen 6.

I have two Halogen projects going at the moment, both of which are complex enough to exercise all of the changes relative to v5.  I've tried patched versions of them against the next-6 branch, and I like the result, especially for simple cases of event sources!  Don't get me wrong: I have been hooked on Halogen ever since first using v0.12 (for a predecessor of one of the present projects), and v5 already made things pretty easy and clear; next-6 just continues to improve it!

Let me know what would help - reviewing documentation? testing? any changes to code?

In any case, thanks for a brilliant piece of software!

### LINKED ISSUE #652: Unkown value H.ComponentDSL
I am getting this error 

These are my imports
```
module Main where

import Prelude
import Affjax as AX
import Affjax.ResponseFormat as ResponseFormat
import Data.Array (mapWithIndex)
import Data.Either (Either(..))
import Data.Maybe (Maybe(..))
import Effect (Effect)
import Effect.Aff.Class (class MonadAff, liftAff)
import Halogen as H
import Halogen.Aff as HA
import Halogen.HTML (ClassName(..))
import Halogen.HTML as HH
import Halogen.HTML.Events as HE
import Halogen.HTML.Properties as HP
import Halogen.Themes.Bootstrap4 as B
import Halogen.VDom.Driver (runUI)
import Simple.JSON as SimpleJSON
```

This is my function
```
eval :: Query ~> H.ComponentDSL ButtonState Query Void m
eval (UpdateButtonState next) = do
  H.modify_ not
  pure next
eval (GetButtonState reply) = do
    reply <$> H.get
```

### PR DESCRIPTION
**Pull request by thomashoneyman**
This PR updates the Halogen guide for Halogen 5. It also:

* Updates the main repository README to link to the new resources, describe Halogen a little more, and point users towards other resources they may find helpful.
* Adds a credit to Matthew Park from Awake Security who designed the new Halogen logo (this will be added to the repository when this PR is merged and the credit is on master)
* Moves the "Changes in v5" document into a dedicated changelog directory so we can release official transition guides for each major library version
* Adds a placeholder in the directory structure for the concepts reference that @garyb has been working on.

This resolves a number of guide-related issues that have cropped up on the repository since the first Halogen 5 release candidate:

* Resolves #633
* Resolves #588 
* Resolves #528 

It also resolves #652 in the sense that the documentation no longer talks about `ComponentDSL` and it's included in the transition guide for Halogen 5.

### Next Steps

This guide has already been through a review by the Halogen maintainers, and now this PR opens it up for some public feedback before it gets merged. However, it's not necessary that the guide is _perfect_ before it gets merged -- just that it correctly introduces the fundamental concepts of Halogen with correct examples and is free of any significant errors.

We can continue iterating on this after it merges; for now we need to ensure that the repository has documentation for Halogen 5 as that's what most everyone is using via package sets these days.

### CODE REVIEW

**Comment by thomashoneyman** [comment]
For reviewers: if you'd like to review the guide, it's much easier to read via previewing the files rendered with GitHub markdown vs. reading the diffs. The guide is revised and expanded from the Halogen 4 guide but little of the content can be compared via a diff.

Also, all the full examples in the repository can now be run via Try PureScript!

File: docs/guide/02-Introducing-Components.md [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+  Increment ->
+    H.modify_ \state -> state + 1
+```
+
+Our `handleAction` function responds to `Decrement` by reducing our state variable by 1, and to `Increment` by increasing our state variable by 1. Halogen provides several update functions you can use in your `handleAction` function; these ones are commonly used:
+
+* `modify` allows you to update the state, given the previous state, returning the new state
+* `modify_` is the same as `modify`, but it doesn't return the new state (thus you don't have to explicitly discard the result, as you would with `modify`)
+* `get` allows you to retrieve the current state
+* `gets` allows you to retrieve the current state and also apply a function to it (most commonly, `_.property` to retrieve a particular field from a record)
```
Comment (JordanMartinez) [inline docs/guide/02-Introducing-Components.md:147]: > most commonly, `_.property` to 

Should this be "`_.fieldName` to retrieve a particular field from a record" ? Or rephrase it to, "`_.property` to retrieve a field name `property` from a record."

File: docs/guide/01-Rendering-Halogen-HTML.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+### `IProp`
+
+When you look up functions from the `Halogen.HTML.Properties` and `Halogen.HTML.Events` modules, you'll see the `IProp` type featured prominently. For example, here's the `placeholder` function which will let you set the string placeholder property on a text field:
+
+```purs
+placeholder :: forall r i. String -> IProp (placeholder :: String | r) i
+placeholder = prop (PropName "placeholder")
+```
+
+The `IProp` type is used to for events and properties. It uses a row type to uniquely identify particular events and properties; when you then use one of these properties with a Halogen HTML element, Halogen is able to verify whether the element you're applying the property to actually supports it.
```
[the remark this answers is not in the document]
Comment (thomashoneyman) [reply docs/guide/01-Rendering-Halogen-HTML.md:187]:
```suggestion
The `IProp` type is used for events and properties. It uses a row type to uniquely identify particular events and properties; when you then use one of these properties with a Halogen HTML element, Halogen is able to verify whether the element you're applying the property to actually supports it.
```

File: docs/guide/02-Introducing-Components.md [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+Just like how our state type had to be paired with an `initialState` function that describes how to produce a `State` value, our `Action` type should be paired with a function called `handleAction` that describes what to do when one of these actions occurs.
+
+```purs
+handleAction :: forall o m. Action -> H.HalogenM State Action () o m Unit
+```
+
+As with our input type, we can leave type variables open for types that we aren't using. 
+
+* The type `()` means our component has no child components. We could also leave it open as a type variable, because we aren't using it, but `()` is so short that you'll see this commonly used.
```
[the remark this answers is not in the document]
Comment (thomashoneyman) [reply docs/guide/02-Introducing-Components.md:124]:
```suggestion
* The type `()` means our component has no child components. We could also leave it open as a type variable because we aren't using it -- `slots`, by convention -- but `()` is so short you'll see this type commonly used instead.
```

File: docs/guide/02-Introducing-Components.md (hunk 1 above)
Comment (thomashoneyman) [reply docs/guide/02-Introducing-Components.md:147]:
```suggestion
* `gets` allows you to retrieve the current state and also apply a function to it (most commonly, `_.fieldName` to retrieve a particular field from a record)
```

**Comment by thomashoneyman** [comment]
@JordanMartinez thanks for the comments! I've fixed things as you've suggested.

**Comment by garyb** [comment]
I looked through the parts you added/revised since I last reviewed and nothing jumped out at me. So if there are no further immediate comments let's geddit innn! :smile: :tada:

**Comment by thomashoneyman** [comment]
@garyb mind if I merge this in 6 hours if there are no further comments?
